### PR TITLE
Reduce f/e bundle size

### LINKED ISSUE #1261: f/e bundle size
## Is your feature request related to a problem? Please describe 👀

On the back of [Jame's issue](https://github.com/openmsupply/health-supply-hub/issues/202) I'd like to double check our f/e bundle size

## Describe the solution you'd like 🎁
Minimise if possible

### PR DESCRIPTION
**Pull request by mark-prins**
Closes #1261

# 👩🏻‍💻 What does this PR do? 
 
Unsure about tree shaking now. Have updated the imports to reduce the larger imports after reading comments around issues with tree shaking in date-fns.
Still irritated by `recharts` and `graphql` both of which are bigger than I think they could be.
The main change is a reduction in `date-fns` due to the locale imports

Some of the size reductions which I noted down - this from the webpack bundle analyzer plugin:

| Package  | Old Size (MB) | New Size (MB ) |
| ------------- | -------------:| -------------:|
| Material UI | 2.11  | 1.72 |
| lodash | 0.724  | 0.193 |
| date-fns | 1.5  | 0.590 |
| **Total** | 10.33 | 8.5 |

The webpack visualizer gives this overview:

**Then**
Screenshot 2023-03-03 at 11 25 06 AM

**Now**
Screenshot 2023-03-03 at 1 14 25 PM

# 🧪 How has/should this change been tested? 

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: client/packages/common/src/authentication/AuthContext.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { IntlUtils } from '@common/intl';
 import { AppRoute } from '@openmsupply-client/config';
 import { LocalStorage, useLocalStorage } from '../localStorage';
 import Cookies from 'js-cookie';
-import { addMinutes } from 'date-fns';
```
Comment (clemens-msupply) [inline client/packages/common/src/authentication/AuthContext.tsx:6]: Oh no, I though a tree shaker is smart enough :(

Comment (mark-prins) [reply client/packages/common/src/authentication/AuthContext.tsx:6]: there were some comments around the `date-fns` package, the way that the modules were exported can cause problems for webpack to implement tree-shaking. It did make a big difference for `lodash`.. so something isn't quite right.

Comment (mark-prins) [reply client/packages/common/src/authentication/AuthContext.tsx:6]: ok. retested and the `date-fns` is able to shake it like a pro.

**Comment by andreievg** [comment]
@mark-prins, I couldn't quite understand why the bundle is still quite large and really wanted to look deeper into this. 

I found that `swc-plugin-transform-import`, wasn't quite being used:

```diff
--- a/client/packages/common/src/hooks/useNativeClient/useNativeClient.ts
+++ b/client/packages/common/src/hooks/useNativeClient/useNativeClient.ts
@@ -1,4 +1,4 @@
-import uniqWith from 'lodash/uniqWith';
+import { uniqWith } from 'lodash';
```
Screenshot 2023-03-06 at 9 45 10 PM

And with the change below, import transform seems to work:

```diff
                 tsx: true,
               },
               target: 'es2015',
-              plugin: m => new PluginTransformImport().visitProgram(m),
             },
+            plugin: (m) => new PluginTransformImport({
+              "lodash": {
+                "transform": "lodash/${member}",
+                "preventFullImport": true
+              }
+            }).visitProgram(m),
           },
         },
         {
```
Screenshot 2023-03-06 at 9 50 35 PM

Next I wanted to apply the same to packages re-export:
```diff
-- a/client/packages/common/src/index.ts
+++ b/client/packages/common/src/index.ts
@@ -11,10 +11,27 @@ export {
   useMatches,
 } from 'kbar';
 
-export * from 'graphql-request';
-export * from 'react-query';
-export * from 'react-query/devtools';
-export * from 'react-router-dom';
+export {
+  useMutation,
+  useQueryClient,
+  useQuery,
+  UseQueryResult,
+  QueryClient,
+  QueryClientProvider,
+} from 'react-query';
+export {
+  matchPath,
+  useLocation,
+  Location,
+  Link,
+  useNavigate,
+  useParams,
+  BrowserRouter,
+  Routes,
+  Route,
+  Navigate,
+  useMatch,
+} from 'react-router-dom';
 export * from './utils';
```

But couldn't quite get `swc-loader` to work, and also noted that it's using `commonjs`:

https://github.com/openmsupply/open-msupply/blob/ff1956d45601a0d7da6aee654e6027cda7cc4eed/client/.swcrc#L12-L14

And from https://webpack.js.org/guides/tree-shaking/:

Screenshot 2023-03-06 at 9 56 49 PM

I tried combination of things to `swc-loader` and import transform plugin to behave, and gave up, last thing was to try `ts-loader`:

**before ts-loader**

Screenshot 2023-03-06 at 10 03 15 PM


**after ts-loader**

much slower (but only for production)

```diff
@@ -57,6 +57,7 @@ module.exports = env => {
     },
     output: {
       publicPath: '/',
+      path: path.resolve(__dirname, 'dist'),
       filename: '[name].[contenthash].js',
       chunkFilename: '[contenthash].js',
       clean: {
@@ -67,24 +68,29 @@ module.exports = env => {
       splitChunks: {
         chunks: 'all',
       },
     },
     module: {
       rules: [
         {
           test: /\.[t|j]sx?$/,
-          loader: 'swc-loader',
+          loader: isProduction ? 'ts-loader' : '
swc-loader',
           exclude: /node_modules/,
-          options: {
-            jsc: {
-              parser: {
-                dynamicImport: true,
-                syntax: 'typescript',
-                tsx: true,
+          options: isProduction
+            ? {
+                /* ts-loader options */
+              }
+            : {
+                /* swc-loader options */
+                jsc: {
+                  parser: {
+                    dynamicImport: true,
+                    syntax: 'typescript',
+                    tsx: true,
+                  },
+                  target: 'es2015',
+                },
               },
-              target: 'es2015',
-              plugin: m => new PluginTransformIm
port().visitProgram(m),
-            },
-          },
         },
         {
           test: /\.css$/,
diff --git a/client/tsconfig.json b/client/tsconf
ig.json
index f95064361..f9c9ee890 100644
--- a/client/tsconfig.json
+++ b/client/tsconfig.json
@@ -13,8 +13,9 @@
     },
     "resolveJsonModule": true,
     "target": "es5",
-    "module": "commonjs",
+    "module": "esnext",
     "jsx": "react",
+    "moduleResolution": "node",
 
     "sourceMap": true,
     "outDir": "./dist",
@@ -28,8 +29,6 @@
     "noImplicitThis": true,
     "alwaysStrict": true,
 
-    "noEmit": true,
-
     "noUnusedLocals": true,
     "noUnusedParameters": true,
--- a/client/package.json
+++ b/client/package.json
@@ -12,7 +12,7 @@
     "build-stats": "lerna run --scope @openmsupply-client/* build-stats",
     "serve": "lerna run --scope @openmsupply-client/* --parallel serve",
     "clean": "lerna run --scope @openmsupply-client/* --parallel clean",
-    "compile": "lerna run --scope @openmsupply-client/* --parallel tsc --since HEAD",
+    "compile": "lerna run --scope @openmsupply-client/* --parallel tsc --since HEAD -- --noEmit",
```
Screenshot 2023-03-06 at 10 07 05 PM

**after ts-loader and re-export changes**

 ```diff
-- a/client/packages/common/src/index.ts
+++ b/client/packages/common/src/index.ts
@@ -11,10 +11,27 @@ export {
   useMatches,
 } from 'kbar';
 
-export * from 'graphql-request';
-export * from 'react-query';
-export * from 'react-query/devtools';
-export * from 'react-router-dom';
+export {
+  useMutation,
+  useQueryClient,
+  useQuery,
+  UseQueryResult,
+  QueryClient,
+  QueryClientProvider,
+} from 'react-query';
+export {
+  matchPath,
+  useLocation,
+  Location,
+  Link,
+  useNavigate,
+  useParams,
+  BrowserRouter,
+  Routes,
+  Route,
+  Navigate,
+  useMatch,
+} from 'react-router-dom';
 export * from './utils';
```

Screenshot 2023-03-06 at 10 08 54 PM

**Letting typescript compiler know there is not side effects**

Adding `sideEffects: false` to all package.json :

`partial diff:`
```diff
index 879dd5c9c..9f5f865ba 100644
--- a/client/package.json
+++ b/client/package.json
@@ -2,6 +2,7 @@
   "name": "openmsupply-client",
   "//": "Version is shared between client and remote server and is located in repo root package.json",
   "version": "0.1.0",
+  "sideEffects": false,
   "private": true,
   "scripts": {
     "start-local": "lerna run --scope @openmsupply-client/* --parallel start-local",
@@ -111,4 +112,4 @@
   "msw": {
     "workerDirectory": "packages/host/public"
   }
-}
+}
\ No newline at end of file
diff --git a/client/packages/common/package.json b/client/packages/common/package.json
index 30048498f..e19387a52 100644
--- a/client/packages/common/package.json
+++ b/client/packages/common/package.json
@@ -1,6 +1,7 @@
 {
   "name": "@openmsupply-client/common",
   "version": "0.0.1",
+  "sideEffects": false,

```
And adding `usedExports:true` to webpack config
```diff
--- a/client/packages/host/webpack.config.js
+++ b/client/packages/host/webpack.config.js
@@ -25,7 +25,7 @@ module.exports = env => {
nerated.
rated
          */
-        analyzerMode: 'disabled',
+        analyzerMode: 'static',
         generateStatsFile: true,
       })
     : new DummyWebpackPlugin();
@@ -57,6 +57,7 @@ module.exports = env => {
     },
     output: {
       publicPath: '/',
+      path: path.resolve(__dirname, 'dist'),
       filename: '[name].[contenthash].js',
       chunkFilename: '[contenthash].js',
       clean: {
@@ -67,24 +68,29 @@ module.exports = env => {
       splitChunks: {
         chunks: 'all',
       },
+      usedExports: true,
     },
```

Screenshot 2023-03-06 at 10 18 42 PM


**Finally changing to es2015**
```diff
--- a/client/tsconfig.json
+++ b/client/tsconfig.json
@@ -12,9 +12,10 @@
       "@common/utils": ["packages/common/src/utils"]
     },
     "resolveJsonModule": true,
-    "target": "es5",
-    "module": "commonjs",
+    "target": "es2015",
+    "module": "esnext",
     "jsx": "react",
+    "moduleResolution": "node",
```
Screenshot 2023-03-06 at 10 34 27 PM

Which is 629 mb gzipped : D

Just for reference difference in lodash (to the one at the top):

Screenshot 2023-03-06 at 10 34 59 PM

Full diff:

https://github.com/openmsupply/open-msupply/compare/1261-bundle-size...1261-example

**Comment by mark-prins** [comment]
wowsers! thanks @andreievg - a very comprehensive investigation!

I've included all your changes, with one addition:

lodash still isn't playing the tree shaking game, with the swc plugin removed:
Screenshot 2023-03-07 at 4 39 10 PM

So I've updated the other lodash imports that have snuck into `develop` and that brings us down to `2.13 MB`

Screenshot 2023-03-07 at 4 46 58 PM

File: client/packages/host/package.json [hunk 2]
```diff
@@ -25,7 +26,7 @@
     "build": "webpack --env production",
     "build-stats": "webpack --env stats --env production",
     "serve": "serve dist -p 3003",
-    "tsc": "tsc --build"
+    "tsc": "tsc"
```
Comment (mark-prins) [inline client/packages/host/package.json:29]: Required for tsc to run:

Screenshot 2023-03-07 at 4 50 36 PM

File: client/packages/inventory/src/Stocktake/DetailView/Footer/StatusChangeButton.tsx [hunk 3]
```diff
@@ -78,9 +80,9 @@ const useStatusChangeButton = () => {
 
   const mapStructuredErrors = (
     result: Awaited<ReturnType<typeof save>>
-  ): /*error*/ string | /*OK*/ undefined => {
+  ): /* error */ string | /* OK */ undefined => {
```
Comment (mark-prins) [inline client/packages/inventory/src/Stocktake/DetailView/Footer/StatusChangeButton.tsx:83]: my linter doesn't like the `/*comments with no spaces*/` syntax 🤷

File: client/packages/host/package.json (hunk 2 above)
Comment (andreievg) [reply client/packages/host/package.json:29]: 🙏 `yarn compile` did work for me ¯\_(ツ)_/¯

**Review by andreievg — APPROVED** [review]
Thanks @mark-prins 🎉

**Pull request merged**